contracts/contracts/strategies/sonic/SonicValidatorDelegator.sol

```diff
@@ -0,0 +1,354 @@
+        uint256 validatorId,
+        uint256 undelegatedAmount
+    );
+    event Withdrawn(
+        uint256 withdrawId,
+        uint256 validatorId,
+        uint256 undelegatedAmount,
+        uint256 withdrawnAmount
+    );
+    event RegistratorChanged(address indexed newAddress);
+    event SupportedValidator(uint256 indexed validatorId);
+    event UnsupportedValidator(uint256 indexed validatorId);
+    event DefaultValidatorIdChanged(uint256 indexed validatorId);
+
+    /// @dev Throws if called by any account other than the Registrator
+    modifier onlyRegistrator() {
+        require(
+            msg.sender == validatorRegistrator,
+            "Caller is not the Registrator"
+        );
+        _;
+    }
+
+    /// @dev Throws if called by any account other than the Registrator or Strategist
+    modifier onlyRegistratorOrStrategist() {
+        require(
+            msg.sender == validatorRegistrator ||
+                msg.sender == IVault(vaultAddress).strategistAddr(),
+            "Caller is not the Registrator or Strategist"
+        );
+        _;
+    }
+
+    constructor(
+        BaseStrategyConfig memory _baseConfig,
+        address _wrappedSonic,
+        address _sfc
+    ) InitializableAbstractStrategy(_baseConfig) {
+        wrappedSonic = _wrappedSonic;
+        sfc = ISFC(_sfc);
+    }
+
+    function initialize() external virtual onlyGovernor initializer {
+        address[] memory rewardTokens = new address[](0);
+        address[] memory assets = new address[](1);
+        address[] memory pTokens = new address[](1);
+
+        assets[0] = address(wrappedSonic);
+        pTokens[0] = address(platformAddress);
+
+        InitializableAbstractStrategy._initialize(
+            rewardTokens,
+            assets,
+            pTokens
+        );
+    }
+
+    /// @notice Returns the total value of Sonic (S) that is delegated validators.
+    /// Wrapped Sonic (wS) deposits that are still to be delegated and any undelegated amounts
+    /// still pending a withdrawal.
+    /// @param _asset      Address of Wrapped Sonic (wS) token
+    /// @return balance    Total value managed by the strategy
+    function checkBalance(address _asset)
+        external
+        view
+        virtual
+        override
+        returns (uint256 balance)
+    {
+        require(_asset == wrappedSonic, "Unsupported asset");
+
+        // add the Wrapped Sonic (wS) in the strategy from deposits that are still to be delegated
+        // and any undelegated amounts still pending a withdrawal
+        balance =
+            IERC20(wrappedSonic).balanceOf(address(this)) +
+            pendingWithdrawals;
+
+        // For each supported validator, get the staked amount and pending rewards
+        uint256 validatorLen = supportedValidators.length;
+        for (uint256 i = 0; i < validatorLen; i++) {
+            uint256 validator = supportedValidators[i];
+            // Get the staked amount and any pending rewards
```

Comment:
⚪️ This comment duplicates the comment above?

Reply:
fixed: https://github.com/OriginProtocol/origin-dollar/pull/2335/commits/1c099468647785362478aa656c9d55ab4ad93f35